File: starter_kits/CSharp/Halite3/hlt/Command.cs
namespace Halite3.hlt
{
    public class Command
    {
        public readonly string command;

        /// <summary>
        /// Create a new Spawn Ship command
        /// </summary>
        /// <returns>Command("g")</returns>
        public static Command SpawnShip()
        {
            return new Command("g");
        }

        /// <summary>
        /// Create a new Dropoff command
        /// </summary>
        /// <returns>Command("g")</returns>
        public static Command TransformShipIntoDropoffSite(EntityId id)
        {
            return new Command("c " + id);
        }

        /// <summary>
        /// Create a new command for moving a ship in a given direction
        /// </summary>
        /// <param name="id">EntityId of the ship</param>
        /// <param name="direction">Direction to move in</param>
        /// <returns></returns>
        public static Command Move(EntityId id, Direction direction)
        {
            return new Command("m " + id + ' ' + (char)direction);
        }

        public static Command Stay(EntityId id)
        {
            return new Command("s " + id);
        }

        public static Command Attack(EntityId attackerId, EntityId targetId)
        {
            return new Command("a " + attackerId + ' ' + targetId);
        }

        public static Command AttackStructure(EntityId attackerId, PlayerId ownerId, Position targetPos)
        {
            return new Command("x " + attackerId + ' ' + ownerId + ' ' + targetPos.x + ' ' + targetPos.y);
        }

        private Command(string command)
        {
            this.command = command;
        }

        public override bool Equals(object obj)
        {
            if (this == obj)
                return true;
            if (obj == null || this.GetType() != obj.GetType())
                return false;

            Command command1 = (Command)obj;

            return command.Equals(command1.command);
        }

        public override int GetHashCode()
        {
            return command.GetHashCode();
        }
    }
}

File: starter_kits/C++/hlt/command.cpp
#include "command.hpp"

#include <iostream>

constexpr char GENERATE= 'g';
constexpr char CONSTRUCT[] = "c ";
constexpr char MOVE[] = "m ";
constexpr char ATTACK[] = "a ";
constexpr char ATTACK_STRUCTURE[] = "x ";
constexpr char STAY[] = "s ";
constexpr char DEFEND[] = "d ";

hlt::Command hlt::command::spawn_ship() {
    return std::string(1, GENERATE);
}

hlt::Command hlt::command::transform_ship_into_dropoff_site(EntityId id) {
    return CONSTRUCT + std::to_string(id);
}

hlt::Command hlt::command::move(EntityId id, hlt::Direction direction) {
    return MOVE + std::to_string(id) + ' ' + static_cast<char>(direction);
}

hlt::Command hlt::command::stay(EntityId id) {
    return STAY + std::to_string(id);
}

hlt::Command hlt::command::attack(EntityId attacker_id, EntityId target_id) {
    return ATTACK + std::to_string(attacker_id) + ' ' + std::to_string(target_id);
}

hlt::Command hlt::command::attack_structure(EntityId attacker_id, PlayerId owner_id, const Position &target_pos) {
    return ATTACK_STRUCTURE + std::to_string(attacker_id) + ' ' + std::to_string(owner_id) + ' ' +
           std::to_string(target_pos.x) + ' ' + std::to_string(target_pos.y);
}

hlt::Command hlt::command::defend(EntityId id) {
    return DEFEND + std::to_string(id);
}
